File: desktop-tauri/src-tauri/src/media.rs
// Tiny localhost media server: serves PNG/JPG/MP4 under the workspace by
// absolute path, with HTTP range support so MP4 scrubbing works in the webview.
//   GET /media?path=<absolute file path>
use std::fs::File;
use std::io::{Read, Seek, SeekFrom};
use std::sync::Mutex;
use std::thread;

use tauri::State;
use tiny_http::{Header, Response, Server, StatusCode};

#[derive(Default)]
pub struct MediaState {
    port: Mutex<Option<u16>>,
}

fn content_type(path: &str) -> &'static str {
    let p = path.to_ascii_lowercase();
    if p.ends_with(".png") {
        "image/png"
    } else if p.ends_with(".jpg") || p.ends_with(".jpeg") {
        "image/jpeg"
    } else if p.ends_with(".webp") {
        "image/webp"
    } else if p.ends_with(".gif") {
        "image/gif"
    } else if p.ends_with(".mp4") || p.ends_with(".m4v") {
        "video/mp4"
    } else if p.ends_with(".webm") {
        "video/webm"
    } else if p.ends_with(".wav") {
        "audio/wav"
    } else if p.ends_with(".mp3") {
        "audio/mpeg"
    } else {
        "application/octet-stream"
    }
}

fn query_path(url: &str) -> Option<String> {
    // url like /media?path=<percent-encoded abs>
    let q = url.split_once('?')?.1;
    for kv in q.split('&') {
        if let Some(v) = kv.strip_prefix("path=") {
            return Some(percent_decode(v));
        }
    }
    None
}

fn percent_decode(s: &str) -> String {
    let bytes = s.as_bytes();
    let mut out = Vec::with_capacity(bytes.len());
    let mut i = 0;
    while i < bytes.len() {
        match bytes[i] {
            b'%' if i + 2 < bytes.len() => {
                let hi = (bytes[i + 1] as char).to_digit(16);
                let lo = (bytes[i + 2] as char).to_digit(16);
                if let (Some(h), Some(l)) = (hi, lo) {
                    out.push((h * 16 + l) as u8);
                    i += 3;
                    continue;
                }
                out.push(bytes[i]);
                i += 1;
            }
            b'+' => {
                out.push(b' ');
                i += 1;
            }
            b => {
                out.push(b);
                i += 1;
            }
        }
    }
    String::from_utf8_lossy(&out).into_owned()
}

fn parse_range(headers: &[Header], len: u64) -> Option<(u64, u64)> {
    for h in headers {
        if h.field.equiv("Range") {
            let v = h.value.as_str();
            if let Some(r) = v.strip_prefix("bytes=") {
                let (a, b) = r.split_once('-')?;
                let start: u64 = a.parse().ok()?;
                let end: u64 = if b.is_empty() { len - 1 } else { b.parse().ok()? };
                if start <= end && end < len {
                    return Some((start, end));
                }
            }
        }
    }
    None
}

fn handle(req: tiny_http::Request) {
    let abs = match query_path(req.url()) {
        Some(p) => p,
        None => {
            let _ = req.respond(Response::from_string("missing path").with_status_code(400));
            return;
        }
    };
    let mut file = match File::open(&abs) {
        Ok(f) => f,
        Err(_) => {
            let _ = req.respond(Response::from_string("not found").with_status_code(404));
            return;
        }
    };
    let len = file.metadata().map(|m| m.len()).unwrap_or(0);
    let ct = Header::from_bytes("Content-Type", content_type(&abs)).unwrap();
    let accept = Header::from_bytes("Accept-Ranges", "bytes").unwrap();

    match parse_range(req.headers(), len) {
        Some((start, end)) => {
            let chunk = end - start + 1;
            if file.seek(SeekFrom::Start(start)).is_err() {
                let _ = req.respond(Response::from_string("seek error").with_status_code(500));
                return;
            }
            let reader = file.take(chunk);
            let cr = Header::from_bytes("Content-Range", format!("bytes {start}-{end}/{len}")).unwrap();
            let resp = Response::new(StatusCode(206), vec![ct, accept, cr], reader, Some(chunk as usize), None);
            let _ = req.respond(resp);
        }
        None => {
            let resp = Response::new(StatusCode(200), vec![ct, accept], file, Some(len as usize), None);
            let _ = req.respond(resp);
        }
    }
}

impl MediaState {
    fn start(&self) -> Result<u16, String> {
        if let Some(p) = *self.port.lock().unwrap() {
            return Ok(p);
        }
        let server = Server::http("127.0.0.1:0").map_err(|e| e.to_string())?;
        let port = server.server_addr().to_ip().map(|a| a.port()).ok_or("no port")?;
        *self.port.lock().unwrap() = Some(port);
        thread::spawn(move || {
            for req in server.incoming_requests() {
                thread::spawn(move || handle(req));
            }
        });
        Ok(port)
    }
}

#[tauri::command]
pub fn start_media(state: State<MediaState>) -> Result<u16, String> {
    state.start()
}

File: apps/desktop/src/main/services/media.ts
import http from 'node:http'
import { createReadStream, existsSync, realpathSync, statSync } from 'node:fs'
import { mkdir, rename, stat, unlink } from 'node:fs/promises'
import { createHash } from 'node:crypto'
import { spawn, type ChildProcess } from 'node:child_process'
import os from 'node:os'
import path from 'node:path'

const MIME: Record<string, string> = {
  png: 'image/png',
  jpg: 'image/jpeg',
  jpeg: 'image/jpeg',
  webp: 'image/webp',
  gif: 'image/gif',
  bmp: 'image/bmp',
  svg: 'image/svg+xml',
  mp4: 'video/mp4',
  m4v: 'video/mp4',
  mov: 'video/quicktime',
  webm: 'video/webm',
  wav: 'audio/wav',
  mp3: 'audio/mpeg',
  m4a: 'audio/mp4',
  aac: 'audio/aac',
  flac: 'audio/flac',
  ogg: 'audio/ogg',
}

/**
 * Localhost static media server with HTTP Range support (video scrubbing).
 * Same URL contract as the Tauri app: GET /media?path=<encoded abs path>&v=…
 * Serving is confined to explicitly allowed roots.
 */
export class MediaService {
  private server: http.Server | null = null
  private port = 0
  private allowedRoots: string[] = []
  private videoJobs = new Map<string, Promise<string>>()
  private resolvedVideos = new Map<string, string>()
  private workers = new Set<ChildProcess>()

  // Roots accumulate as the user switches works; keep a bounded LRU so a long
  // session doesn't grow the allowlist (and the per-request linear scan) forever.
  private static readonly MAX_ALLOWED_ROOTS = 16

  allowRoot(root: string) {
    try {
      const real = realpathSync(root)
      const idx = this.allowedRoots.indexOf(real)
      if (idx !== -1) this.allowedRoots.splice(idx, 1)
      this.allowedRoots.push(real)
      if (this.allowedRoots.length > MediaService.MAX_ALLOWED_ROOTS) this.allowedRoots.shift()
    } catch {
      throw new Error('目录不存在,无法允许媒体访问')
    }
  }

  private isAllowed(p: string): boolean {
    return this.allowedRoots.some((root) => p === root || p.startsWith(root + path.sep))
  }

  async start(): Promise<number> {
    if (this.server) return this.port
    const server = http.createServer((req, res) => {
      void this.handle(req, res).catch((error) => {
        if (res.headersSent) {
          res.destroy(error instanceof Error ? error : undefined)
          return
        }
        res.writeHead(500, { 'Content-Type': 'text/plain; charset=utf-8' })
        res.end(error instanceof Error ? error.message : String(error))
      })
    })
    await new Promise<void>((resolve, reject) => {
      server.once('error', reject)
      server.listen(0, '127.0.0.1', () => resolve())
    })
    this.server = server
    const addr = server.address()
    this.port = typeof addr === 'object' && addr ? addr.port : 0
    return this.port
  }

  private resolveAllowedFile(raw: string): string | null {
    try {
      const real = realpathSync(raw)
      const st = statSync(real)
      if (!st.isFile() || !this.isAllowed(real)) return null
      return real
    } catch {
      return null
    }
  }

  private toolPath(name: 'ffmpeg' | 'ffprobe'): string {
    const envName = name === 'ffmpeg' ? process.env.FFMPEG_PATH : process.env.FFPROBE_PATH
    if (envName) return envName
    const candidates = process.platform === 'win32'
      ? [name]
      : [`/opt/homebrew/bin/${name}`, `/usr/local/bin/${name}`, `/usr/bin/${name}`, name]
    return candidates.find((candidate) => !path.isAbsolute(candidate) || existsSync(candidate)) ?? name
  }

  private runTool(command: string, args: string[], collectStdout = false): Promise<string> {
    return new Promise((resolve, reject) => {
      const worker = spawn(command, args, {
        windowsHide: true,
        stdio: ['ignore', collectStdout ? 'pipe' : 'ignore', 'pipe'],
      })
      this.workers.add(worker)
      let stdout = ''
      let stderr = ''
      worker.stdout?.on('data', (chunk: Buffer) => {
        if (stdout.length < 2_000_000) stdout += chunk.toString('utf8')
      })
      worker.stderr?.on('data', (chunk: Buffer) => {
        stderr = (stderr + chunk.toString('utf8')).slice(-16_000)
      })
      worker.once('error', (error) => {
        this.workers.delete(worker)
        reject(error)
      })
      worker.once('close', (code) => {
        this.workers.delete(worker)
        if (code === 0) resolve(stdout)
        else reject(new Error(stderr.trim() || `${path.basename(command)} exited with code ${code}`))
      })
    })
  }

  private async browserCanPlay(real: string): Promise<boolean> {
    const ext = path.extname(real).slice(1).toLowerCase()
    try {
      const stdout = await this.runTool(
        this.toolPath('ffprobe'),
        ['-v', 'error', '-show_entries', 'stream=codec_type,codec_name,pix_fmt', '-of', 'json', real],
        true,
      )
      const parsed = JSON.parse(stdout) as {
        streams?: Array<{ codec_type?: string; codec_name?: string; pix_fmt?: string }>
      }
      const streams = parsed.streams ?? []
      const video = streams.find((stream) => stream.codec_type === 'video')
      const audio = streams.filter((stream) => stream.codec_type === 'audio')
      if (!video?.codec_name) return false
      if (ext === 'mp4' || ext === 'm4v') {
        const safePixelFormat = !video.pix_fmt || ['yuv420p', 'yuvj420p', 'nv12'].includes(video.pix_fmt)
        return video.codec_name === 'h264' && safePixelFormat && audio.every((stream) =>
          !stream.codec_name || ['aac', 'mp3'].includes(stream.codec_name),
        )
      }
      if (ext === 'webm') {
        return ['vp8', 'vp9', 'av1'].includes(video.codec_name) && audio.every((stream) =>
          !stream.codec_name || ['opus', 'vorbis'].includes(stream.codec_name),
        )
      }
      return false
    } catch {
      return false
    }
  }

  private async compatibleVideo(real: string): Promise<string> {
    const source = await stat(real)
    const key = createHash('sha256')
      .update(`${real}\0${source.size}\0${source.mtimeMs}`)
      .digest('hex')
    const resolved = this.resolvedVideos.get(key)
    if (resolved) return resolved
    const existing = this.videoJobs.get(key)
    if (existing) return existing
    const cacheDir = path.join(os.tmpdir(), 'anime-armory-media-cache')
    const output = path.join(cacheDir, `${key}.mp4`)
    const job = (async () => {
      if (await this.browserCanPlay(real)) return real
      try {
        const cached = await stat(output)
        if (cached.isFile() && cached.size > 0) return output
      } catch {
        // Cache miss.
      }
      await mkdir(cacheDir, { recursive: true })
      const temp = path.join(cacheDir, `${key}.${process.pid}.${Date.now()}.tmp.mp4`)
      try {
        await this.runTool(this.toolPath('ffmpeg'), [
          '-hide_banner', '-loglevel', 'error', '-y', '-i', real,
          '-map', '0:v:0', '-map', '0:a:0?',
          '-c:v', 'libx264', '-preset', 'veryfast', '-crf', '20', '-pix_fmt', 'yuv420p',
          '-c:a', 'aac', '-b:a', '160k', '-movflags', '+faststart', temp,
        ])
        await rename(temp, output)
        return output
      } catch (error) {
        await unlink(temp).catch(() => {})
        throw error
      }
    })()
    this.videoJobs.set(key, job)
    try {
      const result = await job
      this.resolvedVideos.set(key, result)
      while (this.resolvedVideos.size > 64) {
        const oldest = this.resolvedVideos.keys().next().value
        if (!oldest) break
        this.resolvedVideos.delete(oldest)
      }
      return result
    } finally {
      this.videoJobs.delete(key)
    }
  }

  private serve(req: http.IncomingMessage, res: http.ServerResponse, real: string, mime?: string) {
    const size = statSync(real).size
    const ext = path.extname(real).slice(1).toLowerCase()
    const headers: Record<string, string> = {
      'Content-Type': mime ?? MIME[ext] ?? 'application/octet-stream',
      'Accept-Ranges': 'bytes',
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Headers': 'Range',
      'X-Content-Type-Options': 'nosniff',
    }
    const range = req.headers.range
    const m = range ? /^bytes=(\d+)-(\d*)$/.exec(range) : null
    if (m) {
      const start = Number(m[1])
      const end = m[2] ? Math.min(Number(m[2]), size - 1) : size - 1
      if (start >= size || start > end) {
        res.writeHead(416, { 'Content-Range': `bytes */${size}` }).end()
        return
      }
      headers['Content-Range'] = `bytes ${start}-${end}/${size}`
      headers['Content-Length'] = String(end - start + 1)
      res.writeHead(206, headers)
      if (req.method === 'HEAD') res.end()
      else createReadStream(real, { start, end }).pipe(res)
      return
    }
    headers['Content-Length'] = String(size)
    res.writeHead(200, headers)
    if (req.method === 'HEAD') res.end()
    else createReadStream(real).pipe(res)
  }

  private async handle(req: http.IncomingMessage, res: http.ServerResponse) {
    const url = new URL(req.url ?? '/', 'http://127.0.0.1')
    if (url.pathname !== '/media' && url.pathname !== '/video') {
      res.writeHead(404).end()
      return
    }
    const raw = url.searchParams.get('path')
    if (!raw) {
      res.writeHead(400).end('missing path')
      return
    }
    const real = this.resolveAllowedFile(raw)
    if (!real) {
      res.writeHead(404).end()
      return
    }
    if (url.pathname === '/video') {
      const compatible = await this.compatibleVideo(real)
      this.serve(req, res, compatible, compatible === real ? undefined : 'video/mp4')
      return
    }
    this.serve(req, res, real)
  }

  dispose() {
    for (const worker of this.workers) worker.kill()
    this.workers.clear()
    this.videoJobs.clear()
    this.resolvedVideos.clear()
    this.server?.close()
    this.server = null
  }
}
